### PR TITLE
PHP externs in the global namespace

### PR DESCRIPTION
**Pull request by mockey**
Some attempts in ocaml hackery, not ready yet:
Modify gen_member_access in genphp to deal with global namespace
Add metadata @:phpConstants for dealing with constants

### CODE REVIEW

**Comment by mockey** [comment]
I tried to hack something useful into genphp. Can you take a look. if the ocaml is correct like this?

As of now, it's not possible to define externs for functions in the global php namespace without `untyped __php__` and the like. It's possible to add `@:native("")` to a class but the methods are generated with the static accessor `::` then (see #4876). So I modified `gen_member_access` to check for a `cl_path`of `([], "")` and drop the accessor then. Now you can do something like this:
```
@:native("")
extern class Native {
  @:native("get_html_translation_table")
  static function getHtmlTranslationTable(table:TableKind=0, flags:TableFlag=2, encoding:String="UTF-8"):php.NativeArray;
}
```
There are problems withe handling of arguments, though. The version above passes `null` for every missing argument and ignores the defaults. A workaround is:
```
  static inline function getHtmlTranslationTable(table:TableKind=0, flags:TableFlag=2, encoding:String="UTF-8"):php.NativeArray
    return getHtmlTranslationTable(table, flags, encoding);
```
This keeps the default values at least. I guess I'll have to look into that as well...

Another issue are global constants. These are variables without the leading `$`. I added a metadata `@:phpConstants` which can be added to this new `@:enum extern abstract` thing to mark them. So you can write:
```
@:phpConstants
@:native("\\")
@:enum extern abstract TableKind(Int) from Int to Int {
  var HTML_SPECIALCHARS;
  var HTML_ENTITIES;
}
```
Metadata has to be added to ast.ml and to common.ml, right? Is `UsedOn TClass` correct here?
Also I figured out, that it has to be added to `make _decl` in typeload.ml to make it work with abstracts.

Another question: Is it possible somehow, that the compiler sees this as a real onstant? When I try to use this as a default value e.g. I get "Parameter default value should be constant".

File: genphp.ml [hunk 1]
```diff
@@ -781,8 +781,12 @@ and gen_member_access ctx isvar e s =
 		(match !(a.a_status) with
 		| EnumStatics _ ->
 			print ctx "::%s%s" (if isvar then "$" else "") (s_ident s)
-		| Statics _ ->
-			print ctx "::%s%s" (if isvar then "$" else "") (s_ident s)
+		| Statics sta ->
+		  let sep = match sta.cl_path with
```
Comment (Simn) [inline genphp.ml]: Check the indent please.

File: common.ml [hunk 2]
```diff
@@ -465,6 +465,7 @@ module MetaInfo = struct
 		| Op -> ":op",("Declares an abstract field as being an operator overload",[HasParam "The operation";UsedOn TAbstractField])
 		| Optional -> ":optional",("Marks the field of a structure as optional",[UsedOn TClassField])
 		| Overload -> ":overload",("Allows the field to be called with different argument types",[HasParam "Function specification (no expression)";UsedOn TClassField])
+		| PhpConstants -> ":phpConstants",("Marks the fields of an TAbstract as php constants, without $",[Platform Php;UsedOn TClass])
```
Comment (Simn) [inline common.ml]: Maybe I misunderstand the intent, but why is this used on a class instead of on a field?

**Comment by Simn** [comment]
> There are problems withe handling of arguments, though. The version above passes null for every missing argument and ignores the defaults.

Yes that has been a long-term issue. It comes from the fact that PHP has `pf_pad_nulls` (in common.ml) which for normal functions makes sense. It also makes sense for Haxe-externs which in theory could be a thing.

I want to make this respect `@:nativeGen` and/or `@:hxGen`. The question is which one should be the default for externs. Pragmatically I expect Haxe-externs compared to native-externs to be a rarity, so I would say we should _not_ pad `null` on externs unless the extern has `@:hxGen`.

> Metadata has to be added to ast.ml and to common.ml, right?

Right, they are declared in ast.ml and defined in common.ml. 

> Is UsedOn TClass correct here?

For what you did, yes. The question is if that's the best approach (see my line comment).

> Also I figured out, that it has to be added to make _decl in typeload.ml to make it work with abstracts.

Indeed, these are the metadata that are forwarded from the TAbstractDecl to its implementation TClassDecl.

> Another question: Is it possible somehow, that the compiler sees this as a real onstant? When I try to use this as a default value e.g. I get "Parameter default value should be constant".

Not at the moment because `tf_args` uses `tconstant` so there's no way to store anything else.

**Comment by mockey** [comment]
Thanks for looking into it.

> It comes from the fact that PHP has pf_pad_nulls (in common.ml)

Ah, that explains why I couldn't find it in genphp. So it's not only a PHP-issue?

> I would say we should not pad null on externs unless the extern has @:hxGen

Sounds good.

> Maybe I misunderstand the intent, but why is this used on a class instead of on a field

The idea here was to have one class/enum a.k.a `@:enum extern abstract` full of constants. It might be useful to be able to mark single fields. But I'd always prefer a separate enum. Not sure, if that's really the best approach, though.

> tf_args uses tconstant so there's no way to store anything else

Would it be possible to add a value to TConstant that could be used here, like maybe TExtern?

**Comment by Simn** [comment]
Is there any way we could add a test for this?

**Comment by mockey** [comment]
Well, it's very PHP-specific. Basically we could test if the generated expression equals the expression written as `untyped __php__`.  But actually it's about the generated code. Do you have tests already where you test/compare the generated code that comes from using `extern` or `@:native` and the like?

**Comment by Simn** [comment]
Not for PHP (un)fortunately. Alright then, I don't think this could break anything so I'll go ahead and merge it. Thanks!

**Comment by mockey** [comment]
> Not for PHP (un)fortunately.

For other targets there are such tests?

> Alright then, I don't think this could break anything

Most probably not.

> so I'll go ahead and merge it.

Cool, thanks.

Are you planning to work on this `pf_pad_nulls` issue for the next release?

**Pull request merged**